feat: add Dockerfiles and compose file to run locally

avrittrohwer opened the pull request:
This PR addresses #5437 

- [ ] Addresses the issue in full
- [x] Addresses only certain aspects of the issue

## Description

Adds Dockerfiles for API and each site and a docker-compose file to run them all locally.

It also upgrades primsa to ^5.3.0 (which yarn resolved to 5.22.0). This is required to run the api container on mac apple silicon (https://github.com/bloom-housing/bloom/pull/5448#issuecomment-3423815754 with details).  Thank you @YazeedLoonat for the super quick code changes required for the upgrade!

## How Can This Be Tested/Reviewed?

See README.md changes for instructions (if they are not sufficient leave a comment and I'll add more details there).

## Author Checklist:

- [ ] Added QA notes to the issue with applicable URLs
- [ ] Reviewed in a desktop view
- [ ] Reviewed in a mobile view
- [ ] Reviewed considering accessibility
- [ ] Added tests covering the changes
- [ ] Made corresponding changes to the documentation
- [ ] Ran `yarn generate:client` and/or created a migration when required

## Review Process:

- Read and understand the issue
- Ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Either (1) explicitly ask a clarifying question, (2) request changes, or (3) approve the PR, even if there are very small remaining changes, if you don't need to re-review after the updates

emilyjablonski commented on docker-compose.yml:
```diff
@@ -0,0 +1,103 @@
+
+      SHOW_PUBLIC_LOTTERY: "TRUE"
+      SHOW_MANDATED_ACCOUNTS: "FALSE"
+      SHOW_PDWLESS: "FALSE"
```
```suggestion
      SHOW_PWDLESS: "FALSE"
```

emilyjablonski commented on docker-compose.yml:
```diff
@@ -0,0 +1,103 @@
+      - "3002:3002"
+    environment:
+      NODE_ENV: "production"
+      NEXTJS_PORT: 3002
```
This is so unimportant lol but is there a reason this is now 3002 instead of 3000?

avrittrohwer replied:
Oh I missed that, changed to 3000 to match the other scripts

ludtkemorgan commented on README.md:
```diff
@@ -82,6 +82,23 @@ You can also run each process individually from separate terminals with the foll
+Compose are required. The easiest way is to install [Docker
+Desktop](https://docs.docker.com/desktop/).
+
+The database is seeded with the `yarn:dbseed:development` script.
```
suggestion: We typically test with the "staging" data as that one is more realistic. The "dev" data is more randomly generated which we mostly use to test large number of things or try to identify edge cases. "staging" data is more consistent and has all the different scenarios we need to account for.

avrittrohwer commented on docker-compose.yml:
```diff
@@ -0,0 +1,103 @@
+services:
+  db:
+    image: docker.io/postgres:18
+    restart: never
```
Fixed!

avrittrohwer commented on api/Dockerfile.dbseed.dev:
```diff
@@ -0,0 +1,21 @@
+COPY --chown=2002:2002 . .
+RUN yarn prisma generate
+
+CMD [ "yarn", "db:seed:development" ]
```
changed the default to staging

avrittrohwer commented:
> Okay, I think I've addressed all issues except for the issue with the `dbseed` container running on macos apple silicon:
> 
> ```
> dbseed-1  | $ npx prisma db seed -- --environment staging
> dbseed-1  | Running seed command ts-node prisma/seed.ts --environment staging ...
> dbseed-1  | malloc(): unaligned tcache chunk detected
> dbseed-1  | malloc(): unaligned tcache chunk detected
> db-1      | 2025-10-17 22:27:48.752 UTC [110] LOG:  unexpected EOF on client connection with an open transaction
> dbseed-1  |
> dbseed-1  | An error occurred while running the seed command:
> dbseed-1  | Error: Command was killed with SIGABRT (Aborted): ts-node prisma/seed.ts --environment staging
> ```
> 
> I found a few bugs that look similar in prisma's github: https://github.com/prisma/prisma/issues?q=is%3Aissue%20%22malloc()%3A%20unaligned%20tcache%20chunk%20detected%22

Adding `engineType = 'binary'` to https://github.com/bloom-housing/bloom/blob/avritt/localdocker/api/prisma/schema.prisma#L1 eliminated the behavior on my mac.  Looking into https://www.prisma.io/docs/orm/prisma-schema/overview/generators to better understand what this changes.

avrittrohwer commented:
I changed the @prisma/client and prisma dependency requirement to "^5.3.0" in https://github.com/bloom-housing/bloom/blob/avritt/localdocker/api/package.json#L51 and ran `yarn install`.  Yarn resolved it to 5.22.0:

```
diff --git a/api/yarn.lock b/api/yarn.lock
index 2e759a664..3f28c409d 100644
--- a/api/yarn.lock
+++ b/api/yarn.lock
@@ -1885,22 +1885,46 @@
   resolved "https://registry.yarnpkg.com/@pkgjs/parseargs/-/parseargs-0.11.0.tgz#a77ea742fab25775145434eb1d2328cf5013ac33"
   integrity sha512-+1VkjdD0QBLPodGrJUeqarH8VAIvQODIbwh9XpP5Syisf7YoQgsJKPNFoqqLQlu+VQ/tVSshMR6loPMn8U+dPg==

-"@prisma/client@^5.0.0":
-  version "5.0.0"
-  resolved "https://registry.yarnpkg.com/@prisma/client/-/client-5.0.0.tgz#9f0cd4164f4ffddb28bb1811c27eb7fa1e01a087"
-  integrity sha512-XlO5ELNAQ7rV4cXIDJUNBEgdLwX3pjtt9Q/RHqDpGf43szpNJx2hJnggfFs7TKNx0cOFsl6KJCSfqr5duEU/bQ==
-  dependencies:
-    "@prisma/engines-version" "4.17.0-26.6b0aef69b7cdfc787f822ecd7cdc76d5f1991584"
-
-"@prisma/engines-version@4.17.0-26.6b0aef69b7cdfc787f822ecd7cdc76d5f1991584":
-  version "4.17.0-26.6b0aef69b7cdfc787f822ecd7cdc76d5f1991584"
-  resolved "https://registry.yarnpkg.com/@prisma/engines-version/-/engines-version-4.17.0-26.6b0aef69b7cdfc787f822ecd7cdc76d5f1991584.tgz#b36eda5620872d3fac810c302a7e46cf41daa033"
-  integrity sha512-HHiUF6NixsldsP3JROq07TYBLEjXFKr6PdH8H4gK/XAoTmIplOJBCgrIUMrsRAnEuGyRoRLXKXWUb943+PFoKQ==
-
-"@prisma/engines@5.0.0":
-  version "5.0.0"
-  resolved "https://registry.yarnpkg.com/@prisma/engines/-/engines-5.0.0.tgz#5249650eabe77c458c90f2be97d8210353c2e22e"
-  integrity sha512-kyT/8fd0OpWmhAU5YnY7eP31brW1q1YrTGoblWrhQJDiN/1K+Z8S1kylcmtjqx5wsUGcP1HBWutayA/jtyt+sg==
+"@prisma/client@^5.3.0":
+  version "5.22.0"
+  resolved "https://registry.yarnpkg.com/@prisma/client/-/client-5.22.0.tgz#da1ca9c133fbefe89e0da781c75e1c59da5f8802"
+  integrity sha512-M0SVXfyHnQREBKxCgyo7sffrKttwE6R8PMq330MIUF0pTwjUhLbW84pFDlf06B27XyCR++VtjugEnIHdr07SVA==
+
+"@prisma/debug@5.22.0":
+  version "5.22.0"
+  resolved "https://registry.yarnpkg.com/@prisma/debug/-/debug-5.22.0.tgz#58af56ed7f6f313df9fb1042b6224d3174bbf412"
+  integrity sha512-AUt44v3YJeggO2ZU5BkXI7M4hu9BF2zzH2iF2V5pyXT/lRTyWiElZ7It+bRH1EshoMRxHgpYg4VB6rCM+mG5jQ==
+
+"@prisma/engines-version@5.22.0-44.605197351a3c8bdd595af2d2a9bc3025bca48ea2":
+  version "5.22.0-44.605197351a3c8bdd595af2d2a9bc3025bca48ea2"
+  resolved "https://registry.yarnpkg.com/@prisma/engines-version/-/engines-version-5.22.0-44.605197351a3c8bdd595af2d2a9bc3025bca48ea2.tgz#d534dd7235c1ba5a23bacd5b92cc0ca3894c28f4"
+  integrity sha512-2PTmxFR2yHW/eB3uqWtcgRcgAbG1rwG9ZriSvQw+nnb7c4uCr3RAcGMb6/zfE88SKlC1Nj2ziUvc96Z379mHgQ==
``` 

Running `yarn prisma generate` works fine:

```
$ yarn prisma generate
yarn run v1.22.22
$ /home/avritt/github/bloom/api/node_modules/.bin/prisma generate
Prisma schema loaded from prisma/schema.prisma

✔ Generated Prisma Client (v5.22.0) to ./node_modules/@prisma/client in 421ms

Start by importing your Prisma Client (See: https://pris.ly/d/importing-client)

Tip: Want to turn off tips and other hints? https://pris.ly/tip-4-nohints

Done in 1.54s.
```

However now typescript compiler fails:

```
$ yarn build
yarn run v1.22.22
$ rimraf dist
$ nest build
src/services/auth.service.ts:381:5 - error TS2322: Type '{ id: string; passwordHash: string; passwordUpdatedAt: Date; passwordValidForDays: number; resetToken: string; confirmationToken: string; confirmedAt: Date; email: string; firstName: string; ... 17 more ...; wasWarnedOfDeletion: boolean; }' is not assignable to type '{ activityLogs: { id: string; createdAt: Date; updatedAt: Date; userId: string; module: string; action: string; metadata: JsonValue; recordId: string; }[]; applicationFlaggedSet: { ...; }[]; ... 8 more ...; _count: { ...; }; } & { ...; }'.
  Type '{ id: string; passwordHash: string; passwordUpdatedAt: Date; passwordValidForDays: number; resetToken: string; confirmationToken: string; confirmedAt: Date; email: string; firstName: string; ... 17 more ...; wasWarnedOfDeletion: boolean; }' is missing the following properties from type '{ activityLogs: { id: string; createdAt: Date; updatedAt: Date; userId: string; module: string; action: string; metadata: JsonValue; recordId: string; }[]; applicationFlaggedSet: { ...; }[]; ... 8 more ...; _count: { ...; }; }': activityLogs, applicationFlaggedSet, applications, listings, and 7 more.

381     user = await this.prisma.userAccounts.update({
        ~~~~
src/utilities/model-fields.ts:4:3 - error TS4104: The type 'readonly ReadonlyDeep_2<ReadonlyDeep_2<ReadonlyDeep_2<{ kind: FieldKind; name: string; isRequired: boolean; isList: boolean; isUnique: boolean; isId: boolean; isReadOnly: boolean; isGenerated?: boolean; isUpdatedAt?: boolean; type: string; ... 7 more ...; documentation?: string; }>>>[]' is 'readonly' and cannot be assigned to the mutable type 'ReadonlyDeep_2<{ kind: FieldKind; name: string; isRequired: boolean; isList: boolean; isUnique: boolean; isId: boolean; isReadOnly: boolean; isGenerated?: boolean; isUpdatedAt?: boolean; type: string; ... 7 more ...; documentation?: string; }>[]'.

4   return Prisma.dmmf.datamodel.models.filter(
    ~~~~~~

Found 2 error(s).

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

avrittrohwer commented:
Rebuilding all images with `docker compose build` and then running `docker compose up`, both successfully on both my x86 linux host and my apple silicon mac.

The only thing I notice now is that the staging db seed script does not pass in a jurisdiction name (https://github.com/bloom-housing/bloom/blob/94a8e78150dd5b0dc10e29c19abe21caa459624c/api/package.json#L25), so the public site has a lot of ` message: 'jurisdiction Bloomington was requested but not found',` errors

avrittrohwer commented on api/Dockerfile:
```diff
@@ -1,31 +1,53 @@
+USER 2002:2002
+
+# Run any DB migrations then start the server.
+CMD [ "/bin/bash", "-c", "yarn db:migration:run && yarn start:prod" ]
```
@ludtkemorgan I see in your dockerfile PR you had separated out running DB migrations into a different container: https://github.com/bloom-housing/bloom/pull/5170/files#diff-7e3ad9b50cf53c65110682c252aee7766c16b76a8336c41af820c1b27b05628aR63.

Do we want to keep that strategy or are we okay with running DB migrations on the API container startup?

avrittrohwer replied:
If the migration takes longer than 10 seconds, all the containers that are waiting on the lock will exit. For our case AWS will notice the exited container and start a new one to replace it

avrittrohwer commented:
> Rebuilding all images with `docker compose build` and then running `docker compose up`, both successfully on both my x86 linux host and my apple silicon mac.
> 
> The only thing I notice now is that the staging db seed script does not pass in a jurisdiction name (
> 
> https://github.com/bloom-housing/bloom/blob/94a8e78150dd5b0dc10e29c19abe21caa459624c/api/package.json#L25
> 
> ), so the public site has a lot of ` message: 'jurisdiction Bloomington was requested but not found',` errors

Without the explicit jurisdiction name specified in the seed command, a random name is chosen: https://github.com/bloom-housing/bloom/blob/94a8e78150dd5b0dc10e29c19abe21caa459624c/api/prisma/seed-helpers/jurisdiction-factory.ts#L5.  This breaks the public site because it needs the JURISDICTION_NAME env var.  I modified the staging db seed to explicitly use Bloomington so the public site functions as expected.

ludtkemorgan commented on api/package.json:
```diff
@@ -22,7 +22,7 @@
     "db:migration:run": "yarn prisma migrate deploy",
     "db:seed:production": "npx prisma db seed -- --environment production",
-    "db:seed:staging": "npx prisma db seed -- --environment staging",
+    "db:seed:staging": "npx prisma db seed -- --environment staging --jurisdictionName Bloomington",
```
thought: The original reason we didn't add the `jurisdictionName` flag here is because we anticipated using these seeding scripts in our new environments to seed initial data for a new jurisdiction. So we would call this with the new jurisdictionName. Especially with the `production` version of the seeds command

However, I didn't realize that we could still do that because it just takes the last instance of it so 

`yarn db:seed:staging --jurisdictionName newName`

turns into

`npx prisma db seed -- --environment staging --jurisdictionName Bloomington --jurisdictionName newName`

and "newName" is the one chosen. Pretty neat!
